File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {

        if(s.empty()){
            return 0;
        }
        int i=0;
        int sign = 1;
            while(i<s.length() && s[i] == ' '){
            i++;
            }
            if(i>=s.length()){
            return 0;
            }
            if(s[i]=='-'){
                sign = 0;
                i++;
            }
            else if(s[i]=='+'){
                i++;
            }
            long int out = 0;
            while(s[i]>='0' && s[i]<='9'){
                out = out*10;
                if(out<=INT_MIN || out>=INT_MAX){
                    break;
                }
                out += (s[i]-'0');
                i++;
            }
            if(sign==0){
                out *= -1;
            }
            if(out<=INT_MIN){
                out=INT_MIN;
            }
            else if(out>=INT_MAX){
                out = INT_MAX;
            }
        return out;
    }
};

File: 8-string-to-integer-atoi/string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int i = 0;
        int sign = 1;
        int ans = 0;
        // step-1 remove white space
        while(i<s.length() && s.charAt(i) == ' '){
            i++;
        }
        // step-2 handle sign
        if (i < s.length() && (s.charAt(i) == '+' || s.charAt(i) == '-')){
            if(s.charAt(i) == '-'){
                sign = -1;
            }
            else{
                sign = 1;
            }
            i++;
        }
        while(i<s.length() && Character.isDigit(s.charAt(i))){
            int digit = s.charAt(i) - '0';
            if(ans > (Integer.MAX_VALUE - digit)/10){
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            ans = ans * 10 + digit;
            i++;
        }
        return ans*sign;
    }
}